File: app/src/test/java/info/nightscout/androidaps/plugins/constraints/dstHelper/DstHelperPluginTest.java
package info.nightscout.androidaps.plugins.constraints.dstHelper;

import android.content.Context;

import junit.framework.Assert;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

import info.AAPSMocker;
import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.plugins.configBuilder.ConfigBuilderPlugin;
import info.nightscout.androidaps.utils.DateUtil;
import info.nightscout.androidaps.utils.SP;

@RunWith(PowerMockRunner.class)
@PrepareForTest({MainApp.class, ConfigBuilderPlugin.class, SP.class, Context.class})
public class DstHelperPluginTest {
    DstHelperPlugin plugin = new DstHelperPlugin();

    @Test
    public void runTest() throws Exception {
        AAPSMocker.mockMainApp();
        AAPSMocker.mockApplicationContext();

        TimeZone tz = TimeZone.getTimeZone("Europe/Rome");
        TimeZone.setDefault(tz);
        Calendar cal = Calendar.getInstance(tz, Locale.ITALIAN);
        DateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.ITALIAN);
        Date dateBeforeDST = df.parse("2018-03-25 01:55");
        cal.setTime(dateBeforeDST);
        Assert.assertEquals(false, plugin.wasDST(cal));
        Assert.assertEquals(true, plugin.willBeDST(cal));


        TimeZone.setDefault(tz);
        cal = Calendar.getInstance(tz, Locale.ITALIAN);
        dateBeforeDST = df.parse("2018-03-25 03:05");
        cal.setTime(dateBeforeDST);
        Assert.assertEquals(true, plugin.wasDST(cal));
        Assert.assertEquals(false, plugin.willBeDST(cal));


        TimeZone.setDefault(tz);
        cal = Calendar.getInstance(tz, Locale.ITALIAN);
        dateBeforeDST = df.parse("2018-03-25 02:05"); //Cannot happen!!!
        cal.setTime(dateBeforeDST);
        Assert.assertEquals(true, plugin.wasDST(cal));
        Assert.assertEquals(false, plugin.willBeDST(cal));

        TimeZone.setDefault(tz);
        cal = Calendar.getInstance(tz, Locale.ITALIAN);
        dateBeforeDST = df.parse("2018-03-25 05:55"); //Cannot happen!!!
        cal.setTime(dateBeforeDST);
        Assert.assertEquals(true, plugin.wasDST(cal));
        Assert.assertEquals(false, plugin.willBeDST(cal));

        TimeZone.setDefault(tz);
        cal = Calendar.getInstance(tz, Locale.ITALIAN);
        dateBeforeDST = df.parse("2018-03-25 06:05"); //Cannot happen!!!
        cal.setTime(dateBeforeDST);
        Assert.assertEquals(false, plugin.wasDST(cal));
        Assert.assertEquals(false, plugin.willBeDST(cal));

    }

}

File: app/src/test/java/info/nightscout/androidaps/plugins/constraints/dstHelper/DstHelperPluginTest.kt
package info.nightscout.androidaps.plugins.constraints.dstHelper

import dagger.android.AndroidInjector
import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.TestBase
import info.nightscout.androidaps.interfaces.ActivePlugin
import info.nightscout.androidaps.interfaces.Loop
import info.nightscout.androidaps.plugins.bus.RxBus
import info.nightscout.androidaps.utils.resources.ResourceHelper
import info.nightscout.shared.sharedPreferences.SP
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class DstHelperPluginTest : TestBase() {

    @Mock lateinit var rh: ResourceHelper
    @Mock lateinit var sp: SP
    @Mock lateinit var activePlugin: ActivePlugin
    @Mock lateinit var loop: Loop

    private lateinit var plugin: DstHelperPlugin

    val injector = HasAndroidInjector { AndroidInjector { } }

    @Before
    fun mock() {
        plugin = DstHelperPlugin(injector, aapsLogger, RxBus(aapsSchedulers, aapsLogger), rh,
                                 sp, activePlugin, loop)
    }

    @Test
    fun runTest() {
        val tz = TimeZone.getTimeZone("Europe/Rome")
        TimeZone.setDefault(tz)
        var cal = Calendar.getInstance(tz, Locale.ITALIAN)
        val df: DateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.ITALIAN)
        var dateBeforeDST = df.parse("2018-03-25 01:55")
        cal.time = dateBeforeDST!!
        Assert.assertEquals(false, plugin.wasDST(cal))
        Assert.assertEquals(true, plugin.willBeDST(cal))
        TimeZone.setDefault(tz)
        cal = Calendar.getInstance(tz, Locale.ITALIAN)
        dateBeforeDST = df.parse("2018-03-25 03:05")
        cal.time = dateBeforeDST!!
        Assert.assertEquals(true, plugin.wasDST(cal))
        Assert.assertEquals(false, plugin.willBeDST(cal))
        TimeZone.setDefault(tz)
        cal = Calendar.getInstance(tz, Locale.ITALIAN)
        dateBeforeDST = df.parse("2018-03-25 02:05") //Cannot happen!!!
        cal.time = dateBeforeDST!!
        Assert.assertEquals(true, plugin.wasDST(cal))
        Assert.assertEquals(false, plugin.willBeDST(cal))
        TimeZone.setDefault(tz)
        cal = Calendar.getInstance(tz, Locale.ITALIAN)
        dateBeforeDST = df.parse("2018-03-25 05:55") //Cannot happen!!!
        cal.time = dateBeforeDST!!
        Assert.assertEquals(true, plugin.wasDST(cal))
        Assert.assertEquals(false, plugin.willBeDST(cal))
        TimeZone.setDefault(tz)
        cal = Calendar.getInstance(tz, Locale.ITALIAN)
        dateBeforeDST = df.parse("2018-03-25 06:05") //Cannot happen!!!
        cal.time = dateBeforeDST!!
        Assert.assertEquals(false, plugin.wasDST(cal))
        Assert.assertEquals(false, plugin.willBeDST(cal))
    }
}